kong/tools/public.lua

```diff
@@ -12,62 +12,139 @@ do
   local ngx_req_get_post_args = ngx.req.get_post_args
   local ngx_req_get_body_data = ngx.req.get_body_data
 
 
-  local content_type_map = {
-    [1] = function(content_type) -- multipart
-      return multipart(ngx_req_get_body_data(), content_type):get_all()
+  local MIME_TYPES = {
+    form_url_encoded = 1,
+    json             = 2,
+    xml              = 3,
+    multipart        = 4,
+    text             = 5,
+    html             = 6,
+  }
+
+
+  local ERRORS     = {
+    no_ct          = 1,
+    [1]            = "don't know how to parse request body (no Content-Type)",
+    unknown_ct     = 2,
+    [2]            = "don't know how to parse request body (" ..
+                     "unknown Content-Type '%s')",
+    unsupported_ct = 3,
+    [3]            = "don't know how to parse request body (" ..
+                     "can't decode Content-Type '%s')",
+  }
+
+
+  _M.req_mime_types  = MIME_TYPES
+  _M.req_body_errors = ERRORS
+
+
+  local MIME_DECODERS = {
+    [MIME_TYPES.multipart] = function(content_type)
+      local raw_body = ngx_req_get_body_data()
+      local args     = multipart(raw_body, content_type):get_all()
+
+      return args, raw_body
     end,
 
-    [2] = function() -- json
-      local body, err = cjson.decode(ngx_req_get_body_data())
+    [MIME_TYPES.json] = function()
+      local raw_body  = ngx_req_get_body_data()
+      local args, err = cjson.decode(raw_body)
       if err then
         ngx_log(ERR, "could not decode JSON body args: ", err)
-        return {}
+        return {}, raw_body
       end
 
-      return body
+      return args, raw_body
     end,
 
-    [3] = function() -- encoded form
+    [MIME_TYPES.form_url_encoded] = function()
       local ok, res, err = pcall(ngx_req_get_post_args)
       if not ok or err then
         local msg = res and res or err
         ngx_log(ERR, "could not get body args: ", msg)
         return {}
       end
 
+      -- don't read raw_body if not necessary
+      -- if we called get_body_args(), we only want the parsed body
       return res
     end,
   }
 
-  function _M.get_body_args()
+
+  local function get_body_info()
     local content_type = ngx.var.http_content_type
 
     if not content_type or content_type == "" then
-      return {}
+      ngx_log(ERR, ERRORS[ERRORS.no_ct])
+
+      return {}, ERRORS.no_ct
     end
 
-    local map_type
+    local req_mime
 
     if str_find(content_type, "multipart/form-data", nil, true) then
-      map_type = 1
+      req_mime = MIME_TYPES.multipart
 
     elseif str_find(content_type, "application/json", nil, true) then
-      map_type = 2
+      req_mime = MIME_TYPES.json
 
     elseif str_find(content_type, "application/www-form-urlencoded", nil, true) or
       str_find(content_type, "application/x-www-form-urlencoded", nil, true) then
-      map_type = 3
+      req_mime = MIME_TYPES.form_url_encoded
+
+    elseif str_find(content_type, "application/xml", nil, true) then
```

Comment:
I feel like SOAP should be made into its own MIME type (in the above `MIME_TYPE` enum), but let's maybe keep it as XML for now.